### PR TITLE
osd/OSDMap: more pg upmap fixes

### PR DESCRIPTION
**Pull request by xiexingguo**
Fixes: http://tracker.ceph.com/issues/23878

### CODE REVIEW

**Comment by xiexingguo** [comment]
retest this please

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -1623,125 +1623,119 @@ void OSDMap::maybe_remove_pg_upmaps(CephContext *cct,
...
     if (crush_rule < 0) {
       lderr(cct) << __func__ << " unable to load crush-rule of pg "
-                 << p.first << dendl;
+                 << pg << dendl;
       continue;
     }
+    map<int, float> weight_map;
+    auto it = rule_weight_map.find(crush_rule);
+    if (it == rule_weight_map.end()) {
+      auto r = tmpmap.crush->get_rule_weight_osd_map(crush_rule, &weight_map);
```
Comment (hjwsm1989) [inline src/osd/OSDMap.cc]: this way is simple to find whether osd is in bucket

File: src/osd/OSDMap.cc [hunk 2]
```diff
@@ -1623,125 +1623,119 @@ void OSDMap::maybe_remove_pg_upmaps(CephContext *cct,
...
         pending_inc->new_pg_upmap_items.erase(it);
       }
-      if (osdmap.pg_upmap_items.count(p.first)) {
-        pending_inc->old_pg_upmap_items.insert(p.first);
+      if (osdmap.pg_upmap_items.count(pg)) {
+        ldout(cct, 10) << __func__ << " cancel invalid "
+                       << "pg_upmap_items entry "
+                       << "[" << osdmap.pg_upmap.find(pg)->first << ":"
+                       << osdmap.pg_upmap.find(pg)->second << "]"
+                       << dendl;
```
Comment (hjwsm1989) [inline src/osd/OSDMap.cc]: should be osdmap.pg_upmap_items.find(pg)

Comment (xiexingguo) [reply src/osd/OSDMap.cc]: @hjwsm1989 Thank you very much. Fixed!

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-04-28_14:04:47-rados-wip-sage2-testing-2018-04-27-1740-distro-basic-smithi/

- http://tracker.ceph.com/issues/23922

**Pull request merged**